File: Olympiad/COCI/official/2007/contest6_solutions/maraton.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 6 - Task MARATON
  Programming language Pascal
*)

program maraton;

const
   dr : array[1..8] of longint = (-1, -1, 0, 1, 1, 1, 0, -1);
   dc : array[1..8] of longint = (0, 1, 1, 1, 0, -1, -1, -1);

var
   n, i, j, smjer         : longint;
   r1, c1, r2, c2, r3, c3 : longint;
   a                      : array[1..30, 1..30] of char;

begin
   readln(n);
   for i:=1 to n do begin
      for j:=1 to n do
         read(a[i, j]);
      readln;
   end;

   for r1:=1 to n do
      for c1:=1 to n do
         for smjer:=1 to 8 do begin
            if a[r1, c1] = '.' then continue;

            r2 := r1 + dr[smjer]; c2 := c1 + dc[smjer];
            r3 := r2 + dr[smjer]; c3 := c2 + dc[smjer];
            if (r3 < 1) or (r3 > n) then continue;
            if (c3 < 1) or (c3 > n) then continue;
            if (a[r1, c1] = a[r2, c2]) and (a[r2, c2] = a[r3, c3]) then begin
               writeln(a[r1, c1]);
               exit;
            end;
         end;

   writeln('ongoing');
end.

File: Olympiad/COCI/official/2007/contest6_solutions/maraton.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 6 - Task MARATON
  Programming language C++
*/

#include <cstdio>

using namespace std;

int n;
char a[30][30+1];

const int dr[8] = { -1, -1, 0, 1, 1, 1, 0, -1 };
const int dc[8] = { 0, 1, 1, 1, 0, -1, -1, -1 };

int main( void ) {
   scanf( "%d", &n );
   for( int i = 0; i < n; ++i ) scanf( "%s", a[i] );

   try {
      for( int r1 = 0; r1 < n; ++r1 )
         for( int c1 = 0; c1 < n; ++c1 )
            for( int d = 0; d < 8; ++d ) {
               if( a[r1][c1] == '.' ) continue;
               int r2 = r1 + dr[d], c2 = c1 + dc[d];
               int r3 = r2 + dr[d], c3 = c2 + dc[d];
               if( r3 < 0 || r3 >= n ) continue;
               if( c3 < 0 || c3 >= n ) continue;
               if( a[r1][c1] == a[r2][c2] && a[r2][c2] == a[r3][c3] ) throw a[r1][c1];
            }
      printf( "ongoing\n" );
   } catch( char c ) {
      printf( "%c\n", c );
   }
   return 0;
}
